Compatibility with Fabric Transfer API.

gniftygnome opened the pull request:
I'm using Compost in a skyblock mod pack where we have Ductwork and Wooden Hoppers, and I've found a few problems with the modded hoppers.  Basically, Fabric's Transfer API complicates things by hard-coding its own special handler for the Composter.  I discussed that briefly in the Transfer API dev Discord, and the conclusion was at least for now to use a mixin to disable the behavior, so that is what I have done.  After fixing that, there are some issues with insertion that also needed to be resolved, which I did by completing the Inventory implementation and reworking the program flow.

I've tested this with vanilla Hoppers, as well as Ductwork Collectors and Wooden Hoppers and so far I can't find any more problems.  I think this will resolve #4.

- Use a full Inventory implementation
- Disable Fabric Transfer API's special Composter handling
- Move more logic into markDirty()

yurisuika commented:
What issues were you experiencing exactly? I had tested Fabric Wooden Hoppers and did not notice any peculiarities. It figures that Fabric's solution is not needed for a more normal inventory.

One issue here is that the max stack size has to be 64 as that allows users to configure output stack sizes of their choice. Was that specifically for the input stack? Because we can just make setStack return 1 specifically for that 28th slot.

As well, canInsert should only return true if the direction is up. This is to stay consistent with vanilla composter insertion logic.

As for #4, taking these changes to Forge seems to work fine with the vanilla hopper, but it causes Forge Wooden Hopper to crash. Before it would just fill the inventory with the input items (so with 64 slots and 27 stacks that would be a lot just getting stuck in it), which could be extracted via another hopper. If you have no outlet, it would just stop filling when it was full. However, with this change, if there is no outlet, it will then use that 28th input slot and start layering and then crash

The answer to the crash is to ensure that the composter is below level 7 when attempting to invoke addToComposter in the markDirty method, but still the original issue with that mod remains. It seems to be ignoring the canInsert/canExtract logic, the same behavior with what the vanilla hopper was doing before I stopped the Forge insertHook from returning true. It probably has something to do with the Forge's item handler, as the hook does. If that entails rewriting the mod to use Forge item handlers, well it might take some time cause I know nothing about them.

Otherwise, it seems to work the same. I am still curious as to what problems you were experiencing.

gniftygnome commented:
It has been a while; I had to refresh my memory.  The original problem (caused by Fabric transfer API) was the modded hoppers would receive bone meal no matter what was configured.  Disabling the transfer API fixed that, but then of course the modded hoppers could not insert.  I also had timing issues with getting hoppers below the composter to extract.  They would extract one item fine but then they thought the inventory was empty and the composter did not know to send them further block updates.  So I switched to a real inventory to solve those issues; it just seemed better because hoppers understand real inventories without all the special cases.

Yes, I think I had to reduce the max stack size to 1 to prevent queuing items in the composter inventory, which causes various problems.  I could not find any issues with the behavior (I vaguely recall that value is only checked on insertion).  I think making setStack return 1 will still allow hoppers to queue items into the composter when it is waiting to be emptied.  You would probably have to return 0 or 1 depending on whether the input inventory was empty.

I believe I changed the canInsert logic because I liked it better.  Then you can run a double chest of input through a single composter and into a single chest, and it's more compact and tileable.  I don't really see a good reason for the composter to behave differently from every other similar block but it's clearly a preference thing.  So that one is up to you.
